Fixed undefined index errors.

ezzatron opened the pull request:
This PR fixes some undefined index errors that occur when trying to access an undefined scope property.

Currently, if `peridotScanChildren()` finds nothing, it returns `null`, and since it's used like `list($result, $found) = $this->peridotScanChildren(`, this causes undefined index errors.

The tests were passing because an older version of Peridot (before I fixed error handling) gets installed by Composer, and the errors just go off into the ether.

There's a deeper problem here, in that when you run `composer install` on this project, you're getting a very old version of Peridot to test with. I assume this is because `peridot` relies on `peridot-scope`, and to avoid a circular dependency, Composer picks the last version of Peridot that had no dependency on `peridot-scope`. I don't know how best to fix this, other than merging `peridot-scope` back into `peridot`.

brianium commented:
@ezzatron hey man - thanks for all the PRs lately. I have to confess reviewing them has been low on the priority list lately (we don't really have anyone maintaining the org or project right now). Would you at all be interested in taking over? I can give you a run down on things if it helps and give you write access to all the things. If not, I definitely understand :)

ezzatron commented:
Wow, that's a big call; I'm flattered actually.

I'm certainly happy to help out with maintenance if I can. I don't always have a lot of time for open source either, but having write access to this stuff would certainly make it a lot easier to fix these problems that span multiple repos.

How about we have a chat over Hangouts/Skype/whatever? You can fill me in on Peridot/Leo stuff, and I've got a couple of ideas for improvements that are probably easier to explain over voice as well.

brianium commented:
We can definitely do Skype/Hangout/whatever. I am open to anything that works for you. You want to shoot me a DM on twitter (@scaturr) with your details and some times that work? I'm on eastern standard time.

ezzatron commented:
I tried to DM you, but I can't because you don't follow me. I'm on AEST, which is always UTC+10 (we don't observe daylight savings). You're currently on daylight savings, right? If so, then 9pm where you are should be 11am for me, which could work.

I can do any weekday this coming week, except Tuesday. I'm [@ezzatron](https://twitter.com/ezzatron) on Twitter also. Hangouts is probably easiest for me. I'll send you account details once Twitter lets me.

ezzatron committed:
Fixed undefined index errors.

ezzatron commented:
@brianium This PR got a bit bloated unfortunately, but it should be a once-off thing thankfully. 

I'd like to get this released ASAP. I already made a release of `peridot-jumpstart` and it went fine (figured it was pretty hard to screw that one up). I implemented a [changelog](https://github.com/peridot-php/peridot-jumpstart/blob/master/CHANGELOG.md) and [documented the release](https://github.com/peridot-php/peridot-jumpstart/releases/tag/1.3.0), as per my usual routine.

I figure this repo might be a little more complicated. If you're not busy, maybe we could run through the release together?

To explain what happened with this PR in particular, many of Peridot's components have a circular dependency problem. In this case:
- `peridot-scope` dev-depends on `peridot-jumpstart`
- `peridot-jumpstart` depends on `peridot`
- `peridot` depends on `peridot-scope`

When using `composer install`, Composer treats the root package  as `peridot-scope` of an "undefined" version. This is a problem because, starting with [Peridot 1.8.2](https://packagist.org/packages/peridot-php/peridot#1.8.2), there is a requirement for `"peridot-php/peridot-scope": "~1.0"`, which conflicts with this "undefined" version of `peridot-scope` provided by the root package. Therefore Composer finds the latest version of `peridot` that did not require `peridot-scope` at all (1.8.1), and installs that instead of the latest version.

Thankfully, I was able to find a way around this issue. There is an environment variable, [`COMPOSER_ROOT_VERSION`](https://getcomposer.org/doc/03-cli.md#composer-root-version) which allows you to set a temporary root package version without hard-coding it into `composer.json`. So long as you set this to `1.99.99` when using `composer install` or `composer update`, Composer will treat the root project as a compatible version, and hey presto, we get an up-to-date version of Peridot to test with.

This change cascaded on to a few other things. Firstly, remembering to define `COMPOSER_ROOT_VERSION` every time is a pain, so I set up a basic `Makefile` to take care of the heavy lifting. You can now run:
- `make install` to install dependencies
- `make test` to run the tests

Using makefiles sounds kinda old-school, I know, but honestly, `make` is everywhere, and they can be very self-explanatory if kept simple. I've used them to great effect in other projects, and it's something I think that all Peridot components could benefit from.

Other things that flowed on into this PR:
- The `.travis.yml` config was no longer working. I've updated it with some tricks I've learned, and optimized it for build speed. I'd also like to add an encrypted GitHub token to prevent rate-limiting in future, but I can't do that on a PR AFAIK.
- I added `composer.lock` because, amongst other reasons, it will improve build speed. It's not like NPM, adding the lock file will not affect how dependants of `peridot-scope` resolve their own dependencies.
- I added an updated `peridot.php` from the new version of `peridot-jumpstart` I just released. It sets the default spec path, and strict error reporting, which helps with this PR in particular.

So yeah, that's what all the "extra" stuff is here.

brianium commented:
@ezzatron right on man. I'm not opposed to `make` at all. It's a tried and true system.

One of our discussion points was moving `peridot-scope` back into `peridot` proper. Should we consider doing that sooner than later?

This has some implications - most likely with regards to existing plugins. An incremented minor version might alleviate some of that pain?

Sounds like a pain to have to worry about things like `COMPOSER_ROOT_VERSION`, even if we have useful tooling to remove that pain.

Maybe we could do a remote pair session or something to pull scope into the main library. Or start a feature branch and start the discussion there?

ezzatron commented:
I'm kind of in two minds about the whole `peridot-scope` thing now. I think it probably needs to happen at some point, but there are bugs and PRs that we can address right now, that will really help people out. I think they need to take precedence (for now).

Maybe we just ride the circular dependency dragon until 2.0, when we can give scopes a complete rewrite, and be free to make BC breaks as necessary?

brianium commented:
Sounds good to me. Let's ride that dragon for a bit longer.

ezzatron commented:
Okay, awesome man. So, a couple of quick questions:
1.  Do you want to run through a release together, like remote pairing or whatever? Alternatively I can just try making a release here in `peridot-scope`, and you can tell me if there's anything I need to do better next time. I don't mind either way, if you're busy I'm happy to just give it a crack, but I thought I could show you how I normally do a release, and we can discuss releasing with robo, and how to address peridot-php/peridot#184.
2. I noticed that the main repo is using Scrutinizer. I'd love to turn it on across the board. They changed the pricing to free for open source a while back. Can we organize that somehow?

brianium commented:
I love Scrutinizer. I am definitely in favor of adding it everywhere. I think a pair jam sounds good. Want to hit me up on hangouts with a time that works for you? We can share a screen via hangouts or if you have a favorite remote pair tool I would be down for that too.

ezzatron commented:
If it's not too soon for you, we could aim for the same time this week? Maybe even an hour earlier in the day (8PM Thursday for you) if that works better?

Hangouts sounds good. Hopefully it doesn't cut out this time. I'll try to figure out some backup plan for that.

brianium commented:
Tomorrow should be fine. Can we shoot for 9 again though?

ezzatron commented:
Sure, sounds good. Catch you then!
